File: scripts/compile_models.py
#!/usr/bin/env python3
"""
Compile .mlpackage files to .mlmodelc for deployment
"""
import os
import sys
from pathlib import Path
import coremltools as ct

def compile_model(mlpackage_path, output_dir):
    """Compile a single mlpackage to mlmodelc"""
    mlpackage_path = Path(mlpackage_path)
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    
    model_name = mlpackage_path.stem
    output_path = output_dir / f"{model_name}.mlmodelc"
    
    print(f"Compiling {mlpackage_path.name}...")
    try:
        # Load and compile
        model = ct.models.MLModel(str(mlpackage_path))
        model.save(str(output_path))
        print(f"✓ Compiled to {output_path}")
        return True
    except Exception as e:
        print(f"✗ Error compiling {mlpackage_path.name}: {e}")
        return False

def main():
    model_dir = Path("/Users/mac/HeyIm/models/RealisticVision_v51_split-einsum")
    output_dir = model_dir / "compiled"
    
    mlpackages = list(model_dir.glob("*.mlpackage"))
    print(f"Found {len(mlpackages)} models to compile\n")
    
    success_count = 0
    for mlpackage in mlpackages:
        if compile_model(mlpackage, output_dir):
            success_count += 1
        print()
    
    print(f"Compilation complete: {success_count}/{len(mlpackages)} successful")

if __name__ == "__main__":
    main()

File: backend/compile_models.swift
#!/usr/bin/env swift

import Foundation
import CoreML

print("🔨 Pre-compiling Core ML models...")

let basePath = "/Users/mac/HeyIm/models/RealisticVision_v51_split-einsum"
let models = [
    "Stable_Diffusion_version_RealisticVision_v51_diffusers_unet",
    "Stable_Diffusion_version_RealisticVision_v51_diffusers_text_encoder",
    "Stable_Diffusion_version_RealisticVision_v51_diffusers_vae_decoder"
]

for modelName in models {
    let modelPath = "\(basePath)/\(modelName).mlpackage"
    let modelURL = URL(fileURLWithPath: modelPath)
    
    print("\n📦 Compiling: \(modelName)")
    print("   Source: \(modelPath)")
    
    do {
        let compiledURL = try MLModel.compileModel(at: modelURL)
        print("   ✓ Compiled to: \(compiledURL.path)")
        
        // Get size
        if let attrs = try? FileManager.default.attributesOfItem(atPath: compiledURL.path),
           let size = attrs[.size] as? Int64 {
            let sizeMB = Double(size) / 1_000_000.0
            print("   ✓ Size: \(String(format: "%.1f", sizeMB)) MB")
        }
        
    } catch {
        print("   ❌ Failed: \(error)")
    }
}

print("\n✅ Compilation complete!")
